File: tasks/04-sql-reasoning/go/queries_tests.go
// tasks/04‑sql‑reasoning/go/queries_test.go
package queries

import (
	"database/sql"
	"encoding/csv"
	"fmt"
	"os"
	"path/filepath"
	"reflect"
	"testing"

	_ "github.com/mattn/go-sqlite3"
)

func rootDir(t *testing.T) string {
	dir, err := filepath.Abs(filepath.Join(".."))
	if err != nil {
		t.Fatalf("abs path: %v", err)
	}
	return dir
}

func dbPath(t *testing.T) string { return filepath.Join(rootDir(t), "donations.db") }
func expPath(t *testing.T, name string) string {
	return filepath.Join(rootDir(t), "../expected", name)
}

func fetch(t *testing.T, sqlStr string) ([]string, [][]string) {
	db, err := sql.Open("sqlite3", dbPath(t))
	if err != nil {
		t.Fatalf("open db: %v", err)
	}
	defer db.Close()

	rows, err := db.Query(sqlStr)
	if err != nil {
		t.Fatalf("query: %v", err)
	}
	defer rows.Close()

	cols, _ := rows.Columns()
	data := [][]string{}
	for rows.Next() {
		raw := make([]interface{}, len(cols))
		ptrs := make([]interface{}, len(cols))
		for i := range raw {
			ptrs[i] = &raw[i]
		}
		if err := rows.Scan(ptrs...); err != nil {
			t.Fatalf("scan: %v", err)
		}
		rec := make([]string, len(cols))
		for i, v := range raw {
			if v == nil {
				rec[i] = ""
			} else {
				rec[i] = fmt.Sprint(v)
			}
		}
		data = append(data, rec)
	}
	return cols, data
}

func loadExpected(t *testing.T, name string) ([]string, [][]string) {
	f, err := os.Open(expPath(t, name))
	if err != nil {
		t.Fatalf("open exp: %v", err)
	}
	defer f.Close()
	r := csv.NewReader(f)
	header, _ := r.Read()
	body, _ := r.ReadAll()
	return header, body
}

func TestTaskA(t *testing.T) {
	expH, expR := loadExpected(t, "q1.csv")
	gotH, gotR := fetch(t, SQLA)
	if !reflect.DeepEqual(expH, gotH) || !reflect.DeepEqual(expR, gotR) {
		t.Fatalf("Task A mismatch")
	}
}

func TestTaskB(t *testing.T) {
	expH, expR := loadExpected(t, "q2.csv")
	gotH, gotR := fetch(t, SQLB)
	if !reflect.DeepEqual(expH, gotH) || !reflect.DeepEqual(expR, gotR) {
		t.Fatalf("Task B mismatch")
	}
}

File: tasks/04-sql-reasoning/csharp/QueriesTests.cs
// tasks/04‑sql‑reasoning/csharp/QueriesTests.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.IO;
using Microsoft.Data.Sqlite;
using Xunit;

namespace SqlReasoning.Tests
{
    public class QueriesTests
    {
        private static string RootDir => Path.GetFullPath(Path.Combine(AppContext.BaseDirectory, "..", "..", ".."));
        private static string DbPath   => Path.Combine(RootDir, "donations.db");
        private static string ExpDir   => Path.Combine(RootDir, "expected");

        // ---------- helpers ------------------------------------------------ //
        private static (string[] header, List<string[]> rows) Run(string sql)
        {
            using var conn = new SqliteConnection($"Data Source={DbPath}");
            conn.Open();
            using var cmd = conn.CreateCommand();
            cmd.CommandText = sql;

            using var rdr = cmd.ExecuteReader();
            var header = new string[rdr.FieldCount];
            for (int i = 0; i < rdr.FieldCount; i++) header[i] = rdr.GetName(i);

            var rows = new List<string[]>();
            while (rdr.Read())
            {
                var row = new string[rdr.FieldCount];
                for (int i = 0; i < rdr.FieldCount; i++) row[i] = rdr.GetValue(i).ToString();
                rows.Add(row);
            }
            return (header, rows);
        }

        private static (string[] header, List<string[]> rows) LoadCsv(string name)
        {
            var path = Path.Combine(ExpDir, name);
            using var sr = new StreamReader(path);
            var header = sr.ReadLine()?.Split(',');
            var rows   = new List<string[]>();
            string? line;
            while ((line = sr.ReadLine()) != null)
                rows.Add(line.Split(','));
            return (header!, rows);
        }

        // ---------- tests -------------------------------------------------- //
        [Fact]
        public void TaskA_MatchesCsv()
        {
            var exp = LoadCsv("q1.csv");
            var got = Run(Queries.SQL_A);
            Assert.Equal(exp.header, got.header);
            Assert.Equal(exp.rows,   got.rows);
        }

        [Fact]
        public void TaskB_MatchesCsv()
        {
            var exp = LoadCsv("q2.csv");
            var got = Run(Queries.SQL_B);
            Assert.Equal(exp.header, got.header);
            Assert.Equal(exp.rows,   got.rows);
        }
    }
}
